src/kixi/datastore/segmentation/inmemory.clj

```diff
@@ -0,0 +1,170 @@
+        header-line (write-csv [headers])
+        _ (bs/transfer header-line
+                       os
+                       {:close? false})]
+    {:value value
+     :file file
+     :id id
+     :lines 1
+     :size-bytes (alength (.getBytes ^String header-line))
+     :output-stream os}))
+
+(defn index-of
+  [coll value]
+  (first
+   (keep-indexed
+    #(when (= value %2)
+       %1)
+    coll)))
+
+(defn segmentate-file-by-column-values
+  [column-name file]
+  (with-open [contents (io/reader file)]
+    (let [parser (parse-csv contents :strict true)
+          header-line (first parser)
+          dex (index-of header-line column-name)]
+      (if dex
+        (->> (rest parser)
+             (reduce 
+              (fn [value->segment-data line]
+                (let [value (nth line dex)
+                      segment-data (get value->segment-data value
+                                        (new-segment-data header-line value))
+                      line-csv (write-csv [line])]
+                  (bs/transfer line-csv
+                               (:output-stream segment-data)
+                               {:close? false
+                                :append? true})                  
+                  (assoc value->segment-data
+                         value
+                         (-> segment-data
+                             (update :lines
+                                     inc)
+                             (update :size-bytes
+                                     (partial + (alength (.getBytes ^String line-csv))))))))
+              {})
+             vals
+             (map (fn [segment-data]
+                    (.close ^java.io.OutputStream (:output-stream segment-data))
+                    segment-data)))
+        {::seg/reason :invalid-column
+         ::seg/cause column-name}))))
+
+(defn upload-segment
+  [filestore communications]
+  (fn [basemetadata request segment-data]
+    (let [_ (bs/transfer (:file segment-data)
+                         (kdfs/output-stream filestore (:id segment-data)))
+          _ (comms/submit communications 
+                          (assoc (select-keys basemetadata 
+                                              [::ms/type
+                                               ::ms/name
+                                               ::kixi.datastore.schemastore/id]) 
+                                 ::ms/id (:id segment-data)
+                                 ::ms/size-bytes (:size-bytes segment-data)
+                                 ::ms/provanance {::ms/source "segmentation"
+                                                  ::ms/parent-id (::ms/id basemetadata)}
+                                 ::seg/segment {::seg/request request
+                                                ::seg/line-count (:lines segment-data)
+                                                ::seg/value (:value segment-data)}))]
+      segment-data)))
+
+(defmacro while-not->>
```

Comment:
it's definitely that type of macro, but i'm not convinced it's a good idea at all yet :)